kvserver: use vanilla truncation in applySnapshotRaftMuLocked

tbg opened the pull request:
Previously, applySnapshotRaftMuLocked performed the log truncation
that comes with snapshot application in an "ad-hoc" way. In particular,
it wasn't clearing the sideloaded storage.

We know call the same methods {stage,finalize}TruncationRaftMuLocked
that are also called in the regular truncation code, which also clears
the sideloaded storage.

We also ensure that the raft log size is zero and trusted after the
snapshot.

Follow-up to https://github.com/cockroachdb/cockroach/pull/145328.

Epic: CRDB-46488

tbg committed:
kvserver: use vanilla truncation in applySnapshotRaftMuLocked

Previously, applySnapshotRaftMuLocked performed the log truncation
that comes with snapshot application in an "ad-hoc" way. In particular,
it wasn't clearing the sideloaded storage.

We know call the same methods {stage,finalize}TruncationRaftMuLocked
that are also called in the regular truncation code, which also clears
the sideloaded storage.

We also ensure that the raft log size is zero and trusted after the
snapshot.

Epic: CRDB-46488

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145947)
<!-- Reviewable:end -->

pav-kv commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -559,15 +559,26 @@ func (r *Replica) applySnapshotRaftMuLocked(
-	// See: https://github.com/cockroachdb/cockroach/pull/145328#discussion_r2068209588
-	r.store.raftEntryCache.Drop(r.RangeID)
+	// The truncation finalized below, after the snapshot is visible.
+	r.stagePendingTruncationRaftMuLocked(pendingTruncation{
```
The intention of `stage` before the write in case of regular truncations is to trick `Replica.mu`-only readers into seeing a truncated log right away, while `raftMu` section might be doing its writes. They see things consistent: (1) truncated state, (2) last entry ID, (3) entries between the two, (4) log size.

I wonder whether we should put `updateState` here too, since we're changing both so that the log looks empty, and any `Replica.mu` reader coming right after us will not attempt to read entries.

Then, if we throw an append into the mix (i.e. consider that bug case in which `Ready` can return snapshot+entries), it looks like we would want to do this in two stages:

1. set last entry ID == truncated state, and size=0
2. write everything
3. move last entry ID forward, update size to non-zero

pav-kv replied:
`RawNode` sees snapshots first though, this case is luckier than with the truncations. So `RawNode` is already prepared to not read the stale log. But I wonder if we have any `Replica.mu`-only log readers outside `RawNode`. I'd really like to have none.

tbg replied:
>  I'd really like to have none.

I'm not aware of any (based on the callers of `(*replicaLogStorage).entriesLocked` but it's hard to say conclusively.

> I wonder whether we should put updateState here too (and it should be in the same Replica.mu section)

That's a good point. I gave this a try, which raised a question about the semantics of the entry cache.

PTAL

pav-kv commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -505,6 +507,77 @@ func (r *Replica) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
+	//
+	// The truncation finalized below, after the snapshot is visible.
+
+	// TODO(during review): when is it actually safe to clear the cache if we
```
Hmm. So if we clear it here, some `mu` reader can cause it refilled before `r.mu.Lock()` in the next line happens. If we first `stage` and then, after releasing `mu`, clear the cache, we have a situation that entries in the cache are, worst case, in an inconsistent fork.

We are holding `raftMu` here, so no append attempts will happen in the meantime. Any racing reader will observe `RaftTruncatedState == LastEntryID`, and `Term()/Entries()` calls will be short circuited. There will be no access to storage or cache, so both can temporarily contain garbage.

Bottom line: I think it's safe to clear the cache after releasing `mu` but while still holding `raftMu`.

pav-kv commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -505,6 +507,77 @@ func (r *Replica) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
+	// through LoadEntries and via term loading. It would be easier if we only
+	// added to the cache in the one place that seems to matter - entry append.
+	// Is that a change we should make now, while we're here?
+	r.cache.Clear(r.ls.RangeID, math.MaxUint64)
```
Were we not clearing the cache before? Why was it ok?

tbg replied:
we were, here: https://github.com/cockroachdb/cockroach/pull/145947/files#r2088757820

pav-kv replied:
Ah, there was a `Drop()` call. Should we call `Drop()` here instead of `Clear()`? The `math.MaxUint64` smells with an overflow danger, since these low/high index methods tend to have +-1 arithmetics.

tbg replied:
good point, done with next push.

tbg commented on pkg/kv/kvserver/replica_raftstorage.go:
```diff
@@ -584,15 +583,11 @@ func (r *Replica) applySnapshotRaftMuLocked(
-	// accepted the snapshot).
-	//
-	// See: https://github.com/cockroachdb/cockroach/pull/145328#discussion_r2068209588
-	r.store.raftEntryCache.Drop(r.RangeID)
```
@pav-kv

tbg committed:
kvserver: drop cache after apply

tbg committed:
kvserver: address a TODO

tbg committed:
kvserver: rename a field

tbg committed:
kvserver: consolidate raft handling assertion checks

We now

- avoid queueing multiple assertion checks in an apply cycle
- queue the snapshot-induced assertion check at the end, giving us more
  flexibility over when the state is updated (and allowing future refactoring
  that streamlines these updates)

tbg committed:
kvserver: inline one layer of snapshot-specific state updates

This is more straightforward.

tbg committed:
kvserver: inline some no-longer-shared code

tbg commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -505,6 +506,76 @@ func (r *Replica) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
+	//
+	// [1]: this is not properly supported yet and will currently fatal.
+	// See: https://github.com/cockroachdb/cockroach/pull/125530
+	r.stagePendingTruncationSharedRaftMuLockedMuLocked(pendingTruncation{
```
I agree. I did some inlining. See if it is clearer now.

> I think it would be fine if this relation is emphasized in comments or is obvious from lower-level primitives like the fact that both funcs set the RaftTruncatedState.

I'm concretely worried that future changes to the state are not reflected properly in all cases that need to update it. But it's not like the code you've commented on really achieved that.

I did some inlining here - I feel better about where this has landed, PTAL.

Note that I rebased, but the "old" commits were not changed - `kvserver: rename a field` is the first new commit.

tbg committed:
kvserver: avoid leaking sideloaded entries with snapshot

tbg commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -505,6 +506,76 @@ func (r *Replica) stagePendingTruncationRaftMuLocked(pt pendingTruncation) {
+	// The truncation finalized below, after the snapshot is visible.
+
+	// Clear the raft entry cache at the end of this method (after mu has
+	// been released). This means that the cache will lag the disk, but since
```
Comment is updated. I'm still unclear on exactly what the "correct" semantics are. It is hard to reason about readers that hold only `r.mu` for a moment while grabbing log bounds (and then using the cache without synchronization).

pav-kv reviewed (approved):
LGTM % nits. Apply as many as looks reasonable.

tbg committed:
kvserver: comment and style updates
